File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
ListNode *getkth(ListNode *curr, int k)                                                                            
{
    while(curr && k>0)
    {
        curr= curr->next;
        k--;
    }
    return curr;
}
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode * dummy = new ListNode(0,head);
        ListNode * groupprev = dummy;
        while(1)
        {
            ListNode *kth = getkth(groupprev,k);
            if(kth == NULL)
            {
                break;
            }
            ListNode *groupnext = kth->next;
            ListNode * prev = groupnext;
            ListNode *curr = groupprev->next;
            while(curr!=groupnext)
            {
                ListNode * temp = curr->next;
                curr->next = prev;
                prev = curr;
                curr = temp;
            }
            ListNode * temp = groupprev->next;
            groupprev->next = kth;
            groupprev = temp;

        }
        return dummy->next;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseKGroup(self, head, k):
        """
        :type head: Optional[ListNode]
        :type k: int
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(-1,head)
        groupPrev = dummy 
        def getkthnode(node,k):
            curr = node 
            while curr and k>0:
                curr = curr.next
                k-=1
            return curr


        while True:
            kth = getkthnode(groupPrev,k)
            if not kth:
                break
            groupNext = kth.next 
            prev, curr = groupNext, groupPrev.next
            while curr!=groupNext:
                temp = curr.next
                curr.next = prev 
                prev = curr 
                curr = temp
            
            temp = groupPrev.next
            groupPrev.next = prev 
            groupPrev = temp
        return dummy.next 


        